### PR TITLE
Feature/extendend form schema

### LINKED ISSUE #77: How to get labels and hints
A snippet to get labels and hints into the form_schema. Must be improved when translations are present. "guidance" field may need a second look, and an XPATH expert could simplify the paths.

Feel free to use or not

```
library(ruODK)
library(xml2)
library(dplyr)
library(stringr)
library(tidyr)

ru_setup(
  svc = "https:xxxxx",
  un = "yy",
  pw = "xx",
  tz = Sys.timezone()
) 


f_xml = as_xml_document(form_xml())
ff_xml = tibble(
  # Remove trailing /data
  path= str_sub(xml_text(xml_find_all(f_xml, "//translation/text/@id")),6),
  label = xml_text(xml_find_all(f_xml, "//translation/text"))
)%>% 
  separate(path, sep=":", into=c("path", "type")) %>% 
  pivot_wider(names_from=type, values_from=label)

fs_extended = form_schema(flatten = FALSE) %>% 
  left_join(ff_xml, by="path")

```

Here an example from the xml-file
```
        <translation default="true()" lang="Deutsch (de)">
          <text id="/data/pat_gruppe:label">
            <value>Patient</value>
          </text>
          <text id="/data/pat_gruppe/pat_no_barcode:label">
            <value>Patienten-Barcode</value>
          </text>
```

### PR DESCRIPTION
**Pull request by mtyszler**
### Extended Form Schema:

With this pull request I contribute a new function called `form_schema_ext` which has the same function signature as `form_schema`. 

It extends the form schema by adding (multi-language) labels and (multi-language) choice lists from the ODK XML definition.

It closes #77 and addresses the API limitations discussed in https://github.com/getodk/central/issues/172

### CODE REVIEW

**Comment by mtyszler** [comment]
This is still work **Work in Progress**. After I do some stress testing, I'll add proper documentation and follow the test guidelines from [contributing guidelines](https://github.com/ropensci/ruODK/blob/main/CONTRIBUTING.md)

**Comment by florianm** [comment]
Thanks for the PR!
I noticed that the GH actions CI tests all fail due to missing env vars for ru_setup.
If you want to run tests locally, your .Renviron should look like:

```
ODKC_TEST_SVC="https://sandbox.central.getodk.org/v1/projects/14/forms/build_Flora-Quadrat-0-4_1564384341.svc"
ODKC_TEST_URL="https://sandbox.central.getodk.org"
ODKC_TEST_PID=14
ODKC_TEST_FID="build_Flora-Quadrat-0-4_1564384341"
ODKC_TEST_FID_ZIP="build_Spotlighting-0-6_1558333698"
ODKC_TEST_FID_ATT="build_Flora-Quadrat-0-1_1558330379"
ODKC_TEST_FID_GAP="build_Turtle-Track-or-Nest-1-0_1569907666"
ODKC_TEST_FID_WKT="build_Locations_1589344221"
ODKC_TEST_UN="..."
ODKC_TEST_PW="..."
# ODKC_VERSION=0.8
# ODKC_VERSION=0.7
# As long as > 0.7
ODKC_TEST_VERSION=0.8
RU_VERBOSE=TRUE
RU_TIMEZONE="Australia/Perth"
RU_RETRIES=3
```

Would you have a test form with a few translations?

**Comment by mtyszler** [comment]
Hi @florianm , tks for the feedback.

I did have (most of) this locally set-up. I actually used that (as per [contributions guidelines](), with my un and pw):

```
# Required for testing
ODKC_TEST_URL="https://sandbox.central.getodk.org"
ODKC_TEST_PID=14
ODKC_TEST_FID="build_Flora-Quadrat-0-4_1564384341"
ODKC_TEST_FID_ZIP="build_Spotlighting-0-6_1558333698"
ODKC_TEST_FID_ATT="build_Flora-Quadrat-0-1_1558330379"
ODKC_TEST_FID_GAP="build_Turtle-Track-or-Nest-1-0_1569907666"
ODKC_TEST_FID_WKT="build_Locations_1589344221"
ODKC_TEST_UN="your@email.com"
ODKC_TEST_PW="..."
RU_TIMEZOME="Australia/Perth"

```

I did work locally (i.e, it passed the tests locally). I will re-run with the more detailed suggestion you did.  Are you telling me that they will indeed always fail on PR?

There are still a few things I want to add:

1. I want to test with non-default value for parameters (`flatten`, `odata`, `parse`). I might need to restrict this new function for `ODK Central >=0.8` to avoid having to deal with the legacy version. **Let me know if this would be ok** 
2. I want to add a few additional forms for tests
3. Adjust the example in the documentation
4. I'll consider reading the `hints` as well

I do have a small question for you, @florianm. In the new function I use functions from `xml2`. You see I'm using the `xml2::` format for calling those. Let me know if you'd prefer otherwise.

I also got some comments back from `goodpractice` I want to implement.

I'll switch back to ready once I'm done with it.

**Comment by mtyszler** [comment]
Hi @florianm . I think now you can review. I added 3 additional test forms (you can see in the `fid` [here](https://github.com/mtyszler/ruODK/blob/feature/extendend_form_schema/tests/testthat/test-form_schema_ext.R)

Looking forward to hearing from you

File: R/form_schema_ext.R [hunk 1]
```diff
@@ -81,88 +92,29 @@
...
+#' View(fsx)
 #' }
 form_schema_ext <- function(flatten = FALSE, odata = FALSE, parse = TRUE, pid = get_default_pid(),
                             fid = get_default_fid(), url = get_default_url(), un = get_default_un(),
                             pw = get_default_pw(), odkc_version = get_default_odkc_version(),
                             retries = get_retries(), verbose = get_ru_verbose()) {
 
+  # version warning:
+  if (odkc_version < 0.8) {
+    warning("Form Schema Extended works better with ODK Central 0.8 and above",
```
Comment (florianm) [inline R/form_schema_ext.R:113]: For consistency of the console output, I'd use ruODK::ru_msg_warn("...") here

File: R/form_schema_ext.R [hunk 2]
```diff
@@ -193,7 +145,7 @@ form_schema_ext <- function(flatten = FALSE, odata = FALSE, parse = TRUE, pid =
   raw_labels <- xml2::xml_find_all(frm_xml, "//label")
 
   # iterate thorugh labels
```
Comment (florianm) [inline R/form_schema_ext.R:147]: Minor typo ;-)

File: R/form_schema_ext.R [hunk 3]
```diff
@@ -221,38 +173,65 @@ form_schema_ext <- function(flatten = FALSE, odata = FALSE, parse = TRUE, pid =
 
       if (has_translation) {
         # finds all translations related to this path:
-        id <- sub("')", "", sub("jr:itext\\('", "", xml2::xml_attr(this_rawlabel, "ref")))
```
Comment (florianm) [inline R/form_schema_ext.R]: Great to see prefixed function calls. For better or worse, I prefix all external functions, which makes debugging so much easier.

File: R/form_schema_ext.R [hunk 4]
```diff
@@ -264,72 +243,114 @@ form_schema_ext <- function(flatten = FALSE, odata = FALSE, parse = TRUE, pid =
...
+            )
+
+            choice_translations <- all_translations[xml2::xml_attr(
+              all_translations, "id"
+            ) == id_choice]
 
 
             # iterate through choice translations
-            for (kk in 1:length(choice_translations)) {
+            for (kk in seq_along(choice_translations)) {
```
Comment (florianm) [inline R/form_schema_ext.R:304]: Here's a chance to factor out a function and purrr::map() or pmap() over the data, but the code as is with the inline comments reads nicely enough. I'd be happy to keep as is. No need to go overboard with functional pur(rr)itanism here.

**Review by florianm — COMMENTED** [review]
Thanks for the contribution, I've left some comments but this looks good. 
I'll migrate the test forms to a new test server (ETA today if things go to plan) and will push updates to the main branch.
@mtyszler would you have a test form with two or more translations at hand?

**Comment by florianm** [comment]
@mtyszler I've migrated the test server and sent you an invite to the email on your GH profile. Did the email arrive?

With the updated .Renviron settings from the contrib guide and your new credentials you should be able to run the tests locally. Do they work for you? I'm also experimenting with project permissions (you are a "project viewer") to prevent accidental submissions to the test forms (which slows down tests).

I've also added your three test forms for translations and submitted some data, although with shortened form titles and form IDs (apologies for slashing around in your forms). Feel free to add your test cases (or test stubs with test ideas) referencing these forms. You can either reference the form IDs directly or add settings / setters / getters / settings tests for the three I8n forms. The former is probably fine:
```
fid = `Sys.getenv("ODKC_TEST_FID_I8N0", unset="I8n_no_lang")`
fid = `Sys.getenv("ODKC_TEST_FID_I8N1", unset="I8n_label_lng")`
fid = `Sys.getenv("ODKC_TEST_FID_I8N2", unset="I8n_label_choices")`
```

**Comment by mtyszler** [comment]
Thanks @florianm . 

> @mtyszler I've migrated the test server and sent you an invite to the email on your GH profile. Did the email arrive?
> With the updated .Renviron settings from the contrib guide and your new credentials you should be able to run the tests locally. Do they work for you? I'm also experimenting with project permissions (you are a "project viewer") to prevent accidental submissions to the test forms (which slows down tests).

Yes, I did receive the new access. I'll re-run tests locally and let you know .
 
> I've also added your three test forms for translations and submitted some data, although with shortened form titles and form IDs (apologies for slashing around in your forms). Feel free to add your test cases (or test stubs with test ideas) referencing these forms. You can either reference the form IDs directly or add settings / setters / getters / settings tests for the three I8n forms. The former is probably fine:

No worries here... they are for testing purposes, so feel free to arrange as most logical/practical to you. By the way 


> ```
> fid = Sys.getenv("ODKC_TEST_FID_I8N0", unset="I8n_no_lang")
> fid = Sys.getenv("ODKC_TEST_FID_I8N1", unset="I8n_label_lng")
> fid = Sys.getenv("ODKC_TEST_FID_I8N2", unset="I8n_label_choices")
> ```

I'll use these in the tests. I had already included 3 specific tests. I'll see I can add one or two more.

**Comment by mtyszler** [comment]
> Thanks for the contribution, I've left some comments but this looks good.
> I'll migrate the test forms to a new test server (ETA today if things go to plan) and will push updates to the main branch.
> @mtyszler would you have a test form with two or more translations at hand?

@florianm : I'll make sure I pull the main branch in.  One of my current test forms already have English and French. Or do you mean something else?

**Comment by florianm** [comment]
All good - I'm just reviewing the branch - OK if I send a few commits to it? Really nice code btw!
Are you happy for me to add you to the authors as contributor? Have you got an ORCID?

**Comment by mtyszler** [comment]
> All good - I'm just reviewing the branch - OK if I send a few commits to it? Really nice code btw!
> Are you happy for me to add you to the authors as contributor? Have you got an ORCID?
> 
> Edit: forgot I don't have write access to your branch, so seeing that the tests all pass I've merged the PR. So the current main branch should now contain your PR and my latest changes (both related to #96 and minor formatting on this PR).
> Thanks again for the test forms as well - I discovered them on the Sandbox.

Hi @florianm . Tks a lot. I just saw you approved the pull requests and implemented your suggested edits. All great. 

Tks for the quick turn-around. 

If you're ok, I'll delete the branch and fork, as they are not needed anymore.

File: R/form_schema_ext.R (hunk 3 above)
Comment (mtyszler) [reply R/form_schema_ext.R]: I indeed prefer the long way.. it is easier to understand and follow

File: R/form_schema_ext.R (hunk 4 above)
Comment (mtyszler) [reply R/form_schema_ext.R:304]: I indeed prefer the long way.. it is easier to understand and follow

**Comment by florianm** [comment]
Thanks again for the contribution, yep the fork is good to go! I'm waiting for a second PR on encryption and will then release a 1.0 and update the Zenodo release too.

**Pull request merged**